### PR TITLE
make module! arguments optional

### LINKED ISSUE #215: Improve `module` macro
1. Fields should be specifiable in any order
2. Only `license` and `type` should be mandatory, the others should be optional

### PR DESCRIPTION
**Pull request by kloenk**
Fixed: #215

### CODE REVIEW

**Comment by wedsonaf** [comment]
Can you move the parsing of `ModuleInfo` into a function so that we can use it in `module_misc_device` as well?

**Comment by kloenk** [comment]
> Can you move the parsing of `ModuleInfo` into a function so that we can use it in `module_misc_device` as well?

Done, also added alias to the `module_misc_device` and refactored it to use the new `ModuleInfo` struct

File: rust/module.rs [hunk 1]
```diff
@@ -103,6 +103,7 @@ fn expect_end(it: &mut token_stream::IntoIter) {
     }
 }
 
+#[allow(dead_code)]
```
Comment (ojeda) [inline rust/module.rs]: Remove the dead code, if we need them we can reintroduce them later.

File: rust/module.rs [hunk 2]
```diff
@@ -315,6 +325,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
+    author: Option<String>,
+    description: Option<String>,
+    alias: Option<String>,
+    params: Option<Group>,
+}
+
+impl ModuleInfo {
+    fn parse(it: &mut token_stream::IntoIter) -> Self {
+        let mut info = ModuleInfo::default();
+        // the first allowed position is 1, to prevent setting something twice
```
Comment (ojeda) [inline rust/module.rs]: Nit:

```suggestion
        // The first allowed position is 1, to prevent setting something twice.
```

File: samples/rust/rust_chrdev.rs [hunk 3]
```diff
@@ -16,8 +16,7 @@ module! {
     author: b"Rust for Linux Contributors",
     description: b"Rust character device sample",
     license: b"GPL v2",
-    params: {
-    },
```
Comment (ojeda) [inline samples/rust/rust_chrdev.rs:20]: Please remove the empty `params` from the other samples and from binder, since now we can :)

If you move the removals to another commit, that would be ideal.

File: samples/rust/rust_chrdev.rs [hunk 4]
```diff
@@ -16,8 +16,7 @@ module! {
...
+    alias: b"samples_rust_chrdev",
```
Comment (ojeda) [inline samples/rust/rust_chrdev.rs]: Instead of adding an alias to this sample, I would improve the docs of the `module!` macro to include an explanation/example of all the optional keys etc.

Comment (kloenk) [reply samples/rust/rust_chrdev.rs]: Added it to the documentation, could you please check for spelling mistakes :-)

File: rust/module.rs [hunk 5]
```diff
@@ -315,6 +307,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
+            if name == "type" {
+                allowed_pos(&mut pos, 1, &name);
+                let type_ = expect_ident(it);
+                info.type_ = type_;
+            } else if name == "params" {
+                allowed_pos(&mut pos, 7, &name);
+                let params = expect_group(it);
+                info.params = Some(params);
+            } else {
+
```
Comment (ojeda) [inline rust/module.rs]:
```suggestion
```

File: rust/module.rs [hunk 6]
```diff
@@ -365,6 +453,16 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
 ///
+/// # Supported argument types
+///   - `type`: type which implements the [`KernelModule`] trait (required).
+///   - `name`: byte array of the name of the kernel module (required).
+///   - `author`: byte array of the author of the kernel module.
+///   - `description`: byte array of the description of the kernel module.
+///   - `license`: byte array of the license of the kernel module (required).
+///   - `alias`: byte array of alias name of the kernel module.
+///   - `alias_rtnl_link`: byte array of the rtnl_link_alias of the kernel module (mutually exclusive with `alias`).
+///   - `params`: parameters for the kernel module, as decsribed below.
```
Comment (ojeda) [inline rust/module.rs]:
```suggestion
///   - `params`: parameters for the kernel module, as described below.
```

File: rust/module.rs [hunk 7]
```diff
@@ -365,6 +453,16 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
 /// ```
 ///
+/// # Supported argument types
+///   - `type`: type which implements the [`KernelModule`] trait (required).
+///   - `name`: byte array of the name of the kernel module (required).
+///   - `author`: byte array of the author of the kernel module.
+///   - `description`: byte array of the description of the kernel module.
+///   - `license`: byte array of the license of the kernel module (required).
+///   - `alias`: byte array of alias name of the kernel module.
+///   - `alias_rtnl_link`: byte array of the rtnl_link_alias of the kernel module (mutually exclusive with `alias`).
```
Comment (ojeda) [inline rust/module.rs]:
```suggestion
///   - `alias_rtnl_link`: byte array of the `rtnl_link_alias` of the kernel module (mutually exclusive with `alias`).
```

File: rust/module.rs [hunk 8]
```diff
@@ -315,6 +307,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
+                    },
+                    _ => panic!("field '{}' is not supported by module", name),
+                }
+            }
+            assert_eq!(expect_punct(it), ',');
+        }
+
+        expect_end(it);
+
+        info.check_required();
```
Comment (ojeda) [inline rust/module.rs]: Nit: I would paste the definition here and remove the method, since we only call it once, but both ways are fine.

File: rust/module.rs [hunk 9]
```diff
@@ -315,6 +307,102 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
+        }
+    }
+}
+
+fn allowed_pos(pos: &mut u8, allowed_pos: u8, name: &str) {
+    if *pos > allowed_pos {
+        panic!("'{}' is not allowed at this position", name);
+    }
+    *pos = allowed_pos;
+}
```
Comment (ojeda) [inline rust/module.rs]: This one could be inside the `impl`.

Or better, a local lambda so that we only need to write `allowed_pos(x)`.

File: rust/module.rs [hunk 10]
```diff
@@ -746,18 +837,19 @@ pub fn module_misc_device(ts: TokenStream) -> TokenStream {
...
-        type_ = type_,
-        name = name,
-        author = author,
-        description = description,
-        license = license
+        type_ = info.type_,
+        name = info.name,
+        author = info.author.map(|v| format!("author: b\"{}\",", v)).unwrap_or_else(|| "".to_string()),
+        description = info.description.map(|v| format!("description: b\"{}\",", v)).unwrap_or_else(|| "".to_string()),
+        alias = info.alias.map(|v| format!("alias: b\"{}\",", v)).unwrap_or_else(|| "".to_string()),
```
Comment (ojeda) [inline rust/module.rs]: Perhaps we could factor this out, in particular when we get new specialized `module_*!` macros, but it is not a big deal.

File: rust/module.rs (hunk 9 above)
Comment (kloenk) [reply rust/module.rs]: I can't get my head wrapped around making it a lambda at the moment. will put it inside the `impl`.

File: rust/module.rs [hunk 11]
```diff
@@ -315,6 +307,97 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
+        if info.license.is_empty() {
+            panic!("'license' not specified in module macro");
+        }
+        if info.name.is_empty() {
+            panic!("'name' not specified in module macro");
+        }
+
+        info
+    }
+    fn allowed_pos(pos: &mut u8, allowed_pos: u8, name: &str) {
```
Comment (ojeda) [inline rust/module.rs]: Newline between methods/functions please.

File: rust/module.rs [hunk 12]
```diff
@@ -388,178 +481,174 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
-                (
-                    format!("__rust_array_param_{}_{}", vals, max_length),
-                    generated_array_ops_name(vals, max_length),
+    let mut params_modinfo = String::new();
+    if let Some(params) = info.params {
+        assert_eq!(params.delimiter(), Delimiter::Brace);
+
+        let mut it = params.stream().into_iter();
+
+
```
Comment (ojeda) [inline rust/module.rs]: Double empty line.

File: rust/module.rs [hunk 13]
```diff
@@ -315,6 +307,97 @@ fn generated_array_ops_name(vals: &str, max_length: usize) -> String {
...
+    author: Option<String>,
+    description: Option<String>,
+    alias: Option<String>,
+    params: Option<Group>,
+}
+
+impl ModuleInfo {
+    fn parse(it: &mut token_stream::IntoIter) -> Self {
+        let mut info = ModuleInfo::default();
+        // The first allowed position is 1, to prevent setting something twice
```
Comment (ojeda) [inline rust/module.rs]: We are ending the comments with a period (I am documenting this).

**Review by ojeda — CHANGES_REQUESTED** [review]
Since you have to fix the `rustfmt` formatting to pass the CI, I took the change to add a couple more comments...

File: rust/module.rs (hunk 11 above)
Comment (kloenk) [reply rust/module.rs]: I'm sorry, just copied it around. Also rustfmt(check) does not work with O=build. did not see that until now.

**Review by ojeda — APPROVED** [review]

File: rust/module.rs (hunk 11 above)
Comment (ojeda) [reply rust/module.rs]: > Also rustfmt(check) does not work with O=build. did not see that until now.

Fixed on #240.
